[work in progress] Backfill component tests & refactor components for central App container

lmorchard opened the pull request:
This is now a combined PR for Issue #1399 & #1364:
1. refactoring components to work with a central App container all Redux & browser API know-how
2. unit tests to assert that all this changed stuff still works

Tests left to backfill: 
- components
  - <strike>ExperimentRowCard</strike> #1541
  - <strike>ExperimentTourDialog</strike> #1542 
  - [x] LandingPage
  - <strike>MainInstallButton</strike> #1543 
  - [x] Restart
  - [x] RetireConfirmationDialog
  - [x] RetirePage
  - [x] SharePage
- reducers - all, may skip
- actions - all, may skip

lmorchard commented on frontend/test/app/components/NotFoundPage-test.js:
```diff
@@ -0,0 +1,14 @@
+describe('app/components/NotFoundPage', () => {
+  it('should render notFoundHeader string', () => {
+    expect(shallow(<NotFoundPage />)
+      // HACK: .find('[data-l10n-id="errorMessage"]') seems not to work
```
Seems like attribute names with dashes make enzyme unhappy. Maybe better to actually add some descriptive classes to these elements as test handles?

lmorchard commented on frontend/test/app/components/EmailDialog-test.js:
```diff
@@ -0,0 +1,88 @@
+  };
+
+  beforeEach(function() {
+    global.window = {
```
Hmm this window / ga stuff is coming up a lot, I should refactor it into a shared util if I can

lmorchard committed:
Consolidate containers, refactor components

- Merge all containers into single central App container

- Move all code dealing with browser & Redux APIs into helper function
  injected into components from the App container

  - browser e.g. window.location, window.open, cookies, scroll handling

- Move code dealing with external services into App container

  - i.e. sendToGA(),subscribeToBasket()

- Remove sendToGA & subscribeToBasket from lib/utils

- Ensure all components have propTypes declared

- Convert all stateless function components to proper React.Component
  subclasses

- Move Restart from containers to components

- Tweak installAddon in lib/addon to use injected requireRestart helper,
  rather than raw reference to Redux store

- Moved some logic from component constructors to render() in order to
  react better to prop changes - i.e. from tests

- Pass `{...this.props}` along to most sub-components from page
  compoents so that things injected from App get passed along

- When experiment is not yet launched, just render 404 in place rather
  than redirect to /not-found

- Bugfix in Header where dialogs appearing would shift the site
  logo over to the middle of the page

Fixes #1364

lmorchard committed:
Backfill component tests

- Adding tests for components

- Updated test runner command to use frontend/test-setup.js

- New dependencies: jsdom, sinon

Issue #1399

dannycoates commented on frontend/src/app/components/ExperimentPage.js:
```diff
@@ -48,14 +48,15 @@ export default class ExperimentPage extends React.Component {
 
-    const nextExperiment = getExperimentBySlug(nextProps.params.slug);
+    const { isExperimentEnabled: nextIsExperimentEnabled,
+            getExperimentBySlug: nextGetExperimentBySlug } = nextProps;
```
👀 I did a double take when I saw this. To me `nextProps.getExperimentBySlug(...)` is more readable.

dannycoates replied:
actually, there's something unsettling to me about having these functions as props at all. I haven't dug into it but it _feels_ like this component should have an `experiment` prop instead. anyway... maybe a change for another day

lmorchard replied:
Yeah, the only reason it doesn't have an `experiment` prop is because it only gets a `params` prop from the router and has to look it up from the list of available experiments. There could be a better way to do that.

This whole function is also kind of a big hack in that it's tracking experiment enable state changes so we can schedule post-install things like showing a tour dialog and ensuring the button states are turned off. Would be great to rework all of this so it's stateless yet just does the right thing.

But, having these functions as props is what I read as the suggested way of injecting external dependencies into components in the React + Redux pattern. The complication I realized while writing tests is that [these functions are closures around a particular Redux state instance in the `connect()` call in `App`](https://github.com/lmorchard/testpilot/blob/e94db6b230f985ecf518a951c854b0f7660c220c/frontend/src/app/containers/App.js#L62). So, when the Redux state updates, `connect()` updates the props with fresh closures for these helper functions. Using the old functions would mean operating on the previous state.

lmorchard replied:
Oh, but I think I will back out this destructuring magic to be more verbose with the `nextProps` reference. Probably trying to be too clever, here

dannycoates commented on frontend/src/app/components/ExperimentPage.js:
```diff
@@ -141,13 +149,13 @@ export default class ExperimentPage extends React.Component {
             introduction, measurements, privacy_notice_url, changelog_url,
             thumbnail, survey_url, contributors, details, min_release } = experiment;
 
+    // TODO: #1138 - add optional subtitles the right way
```
this should be easy now, right? can we Just Do It®?

lmorchard committed:
Backfill test for RetirePage component

- Move hack to set window.navigator.testpilotAddon on fake uninstall
  into a helper injected from App

- Move FAKE_UNINSTALL_DELAY to props.fakeUninstallDelay in RetirePage

- Add prop types & default props to RetirePage

- Tests

Issue #1399

lmorchard committed:
Backfill tests for SharePage component

- Add prop types to SharePage

- Tests

Issue #1399

lmorchard committed:
Backfill tests for Restart page

- Add propTypes to Restart component

- Tests

Issue #1399

lmorchard committed:
Backfill tests for RetireConfirmationDialog

- Move uninstallAddon to helper injected by App container

- Add propTypes to RetireConfirmationDialog

- Tests

Issue #1399

lmorchard committed:
Add propTypes to ExperimentListPage

lmorchard committed:
Backfill tests for LandingPage

Issue #1399

lmorchard committed:
Simplify nextProps handling in ExperimentPage component mount

lmorchard committed:
Add uninstallAddon to propTypes in Header and ExperimentPage

lmorchard committed:
Add uninstallAddon to other component tests to satisfy warnings

dannycoates commented on frontend/src/app/components/Header.js:
```diff
@@ -152,6 +153,9 @@ export default class Header extends React.Component {
+  sendToGA: React.PropTypes.func.isRequired,
+  openWindow: React.PropTypes.func.isRequired,
   hasAddon: React.PropTypes.bool,
   forceHideSettings: React.PropTypes.bool
```
`forceHideSettings` looks orphaned now; nothing sets it. I'll add a todo to delete it.

lmorchard replied:
Oh, actually that should be used in the Restart component!

lmorchard replied:
[Added a commit to fix that](https://github.com/mozilla/testpilot/pull/1488/files#diff-014a13e8ad141df734228aa1890d34e8L38).

Missed writing a test for that, because I'm kind of confused [what this bit does](https://github.com/lmorchard/testpilot/blob/4f5bfc1f9da345359b4945d7a656b23df10b7bd9/frontend/src/app/components/Header.js#L22) and then forgot about it.

lmorchard committed:
Re-add forceHideSettings to Header in Restart

dannycoates reviewed (approved):
👏
